### ISSUE TITLE
Define "monetary base" and "audit" in the context of Zcash

### ISSUE THREAD

**Issue by Austin-Williams** [zcash/zcash#2289]
For users to have confidence in Zcash as a long term store of value, they must have a way of verifying that the currency is not being fraudulently inflated (via a bug in the verifier, an attacker in possession of toxic waste, etc).

It would behoove us to:

1. Formally define exactly what we mean by "monetary base" and "audit" in the context of Zcash.
2. Determine whether it is possible for users to audit the Zcash monetary base.
3. In the event that an audit is not possible, consider ways to detect and/or mitigate against the fraudulent inflation of the monetary base that do not rely solely on the verifier (such as Zooko's "turnstiles").

It would be, I think, overly ambitious to tackle all three points in a single "issue". However, formalizing what we mean by "monetary base" and "audit" in the context of Zcash will help us tackle the remaining problems, and that's the focus of this issue.

In the context of Bitcoin, it is clear what we mean by "monetary base" and "audit". The sum of the values held by the UTXOs serves as a good definition of "monetary base". Those values are available to all users. An "audit" amounts to summing those public UTXO values and comparing that sum against the sum expected according to Bitcoin's scheduled issuance.

As soon as we start cryptographically hiding values in transactions, though, it becomes less clear what _exactly_ we mean by "monetary base" and -- especially -- "audit".

**Comment by daira** [comment]
To summarise the [Twitter thread](https://mobile.twitter.com/onewayfunction/status/855202193569337344):

A plausible candidate definition (from @nathan-at-least) is "the sum of UXTO [values] plus the secret values of committed unspent notes". Note that this definition applies to the block chain at (before or after) a given block.

The difficulty, as @Austin-Williams pointed out, is that there can be multiple openings of a commitment; the security definition of binding for a commitment scheme only requires it to be *infeasible to find* multiple openings, not that they don't exist. This difficulty applies to both note commitments (in determining which secret values are committed) and nullifiers (in determining which committed notes have been spent). In the latter case, the definition of "spent" is that the nullifier defined as PRF<sup>nf</sup><sub>a<sub>sk</sub></sub>(ρ) has been revealed, where a<sub>pk</sub> = PRF<sup>addr</sup><sub>a<sub>sk</sub></sub>(0) and ρ are from the committed note. So multiple solutions for any of the note fields or multiple preimages for PRF<sup>addr</sup> can cause problems.

In practice some (maybe all) of the undesired solutions will be ruled out by inconsistency with the corresponding proofs, since each proof is done for a specific witness, not just any valid witness. However, we don't want to rely on the proofs for the purpose of defining monetary base; it's too complicated and would not rule out weaknesses in the proving system affecting the audit result.

Note that collisions in the Merkle tree over note commitments do not affect the definition; the set of note commitments and nullifiers is well-defined because they are publicly revealed, and the Merkle tree could always be recomputed with a different hash function. (These do affect the validity of proofs, which is another reason not to make the definition dependent on proofs.)

**Comment by daira** [comment]
It may be less satisfactory, but we could define auditing via an adversarial game. That is, if I am an auditor then I can demonstrate failure of monetary balance by *finding* notes and corresponding a<sub>sk</sub> values such that the resulting monetary base is not as expected. (This is a simplification &ndash;since it only considers overall monetary base&ndash; of the definition of Balance in the Zerocash paper.) But demonstrating that monetary balance holds is then proving a negative, so this definition probably wouldn't be satisfactory to skeptics.

**Comment by daira** [comment]
How about this: a "conditional monetary balance proof" is a proof that there exist openings of all the note commitments and preimages of PRF<sup>addr</sup> such that the monetary balance property holds for those openings and preimages.

Then given a conditional monetary base proof, we would only need to believe that the note commitment scheme is hiding and that PRF<sup>addr</sup> is second-preimage resistant, in order to infer that monetary balance is preserved.

The turnstile method can be viewed as an indirect implementation of this because it does require proofs of note commitment openings and PRF<sup>addr</sup> preimages for all notes up to the turnstile.

**Comment by tromer** [comment]
The property of "total spendable money never exceeds the total that was created" is a special case of the "Balance" security property defined in the [extended version of the Zerocash paper](http://zerocash-project.org/media/pdf/zerocash-extended-20140518.pdf), Appendix C.3 (with the trivial adaptation from the change from Mint+Pour to JoinSplit).

Appendix D.3 proves that this property indeed holds for the Zerocash protocol (and thus, with trivial adaptations, the Zcash protocol).

So in principle, the notion of "auditing the Zcash monetary base" is a red herring. You "audit" a property that could go wrong if someone misbehaves, to make sure they didn't err or cheat. But here we already have a proof that the property holds.

Another way to phase it is that the Zcash protocol (and in particular its zk-SNARK proofs) already do the auditing, on-the-fly.

Of course, in practice there could be mistakes in the proof, adaptation to Zcash, implementation, or parameter generation ceremony. Or the cryptographic assumptions may be broken. So what you're _really_ asking for is another, *redundant* way to be check the same property. And to make it useful, it needs to be sufficiently different from the existing proof so that they wouldn't have common failure modes. Which makes sense, but let's be clear about it and in particular not play along with the silly narrative that "Zcash is not auditable".

**Comment by daira** [comment]
@tromer wrote:
> Appendix D.3 proves that this property indeed holds for the Zerocash protocol (and thus, with trivial adaptations, the Zcash protocol).

Apart from the errors in that appendix pointed out in section 8.8 (numbering as in the Sapling version) of the protocol spec, yes. Which may seem like nitpicking, but it demonstrates the point: this proof depends on a lot of assumptions, and errors are entirely plausible. When we talk about auditing, we *mean* auditing independent of the JoinSplit proofs, and I hope that is clear to everyone reading this thread since it has been explicitly stated in *every* comment above.

**Comment by ioptio** [comment]
Wondering if the Sapling turnstile documentation satisfies this enough to close https://zcash.readthedocs.io/en/latest/rtd_pages/sapling_turnstile.html or are we looking for formal definitions via the spec?

**Comment by ioptio** [comment]
I believe this ticket has been satisfied per above.

**Comment by daira** [comment]
I don't think it has. There's a formal issue of how to define monetary base independent of the proof system, which has not been resolved (and would be useful to resolve).
